File: python/affine_cipher.py
import math


def encode(plain_text, a, b):
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")

    s = ""
    ct = 0
    for c in plain_text:
        if c.isdigit():
            t = c
        else:
            if not c.isalpha():
                continue
            i = ord(c.lower()) - ord("a")
            t = chr((a * i + b) % 26 + ord("a"))

        if ct > 0 and ct % 5 == 0:
            s += " "
        s += t
        ct += 1

    return s


def mmi(a, m):
    for x in range(m):
        if a * x % m == 1:
            return x


def decode(ciphered_text, a, b):
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")

    s = ""
    for c in ciphered_text:
        if c == " ":
            continue
        if c.isdigit():
            s += c
        else:
            y = ord(c) - ord("a")
            s += chr(mmi(a, 26) * (y - b) % 26 + ord("a"))
    return s

File: elixir/affine_cipher.ex
defmodule AffineCipher do
  @m 26

  @typedoc """
  A type for the encryption key
  """
  @type key() :: %{a: integer, b: integer}

  @doc """
  Encode an encrypted message using a key
  """
  @spec encode(key :: key(), message :: String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def encode(%{a: a, b: b}, message) do
    if Integer.gcd(a, @m) > 1 do
      {:error, "a and m must be coprime."}
    else
      message
      |> String.downcase()
      |> String.replace(~r/[^a-z0-9]/, "")
      |> String.to_charlist()
      |> Enum.map(fn
        c when c in ?0..?9 -> c
        c -> rem(a * (c - ?a) + b, @m) + ?a
      end)
      |> Enum.chunk_every(5)
      |> Enum.map(&List.to_string/1)
      |> Enum.join(" ")
      |> then(&{:ok, &1})
    end
  end

  @doc """
  Decode an encrypted message using a key
  """
  @spec decode(key :: key(), message :: String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def decode(%{a: a, b: b}, encrypted) do
    if Integer.gcd(a, @m) > 1 do
      {:error, "a and m must be coprime."}
    else
      mmi = Enum.find(1..25, fn mmi -> rem(a * mmi, @m) == 1 end)

      encrypted
      |> String.replace(" ", "")
      |> String.to_charlist()
      |> Enum.map(fn
        c when c in ?0..?9 -> c
        c -> rem(mmi * Integer.mod(c - ?a - b, @m), @m) + ?a
      end)
      |> List.to_string()
      |> then(&{:ok, &1})
    end
  end
end
